### PR TITLE
Deduplicate instruction prefixes when disassembling

### LINKED ISSUE #42: `flexdis86` gives incorrect size to `xchg ax,ax`'s operands
The output of `utils/dump.sh` on `xchg ax,ax` and `xchg eax,eax` reveals something unusual:

```
$ ./utils/dump.sh xchg ax,ax
66 90                   xchg   ax,ax
II {iiLockPrefix = NoLockPrefix, iiAddrSize = Size64, iiOp = "xchg", iiArgs = [(DWordReg eax,OpType (Opcode_reg 0) VSize),(DWordReg eax,OpType (Reg_fixed 0) VSize)], iiPrefixes = Prefixes {_prLockPrefix = NoLockPrefix, _prSP = SegmentPrefix {unwrapSegmentPrefix = 0}, _prREX = 0b00000000, _prVEX = Nothing, _prASO = False, _prOSO = False}, iiRequiredPrefix = Nothing, iiOpcode = [102,144], iiRequiredMod = Nothing, iiRequiredReg = Nothing, iiRequiredRM = Nothing}
nop

$ ./utils/dump.sh xchg eax,eax
87 c0                   xchg   eax,eax
II {iiLockPrefix = NoLockPrefix, iiAddrSize = Size64, iiOp = "xchg", iiArgs = [(DWordReg eax,OpType ModRM_rm VSize),(DWordReg eax,OpType ModRM_reg VSize)], iiPrefixes = Prefixes {_prLockPrefix = NoLockPrefix, _prSP = SegmentPrefix {unwrapSegmentPrefix = 0}, _prREX = 0b00000000, _prVEX = Nothing, _prASO = False, _prOSO = False}, iiRequiredPrefix = Nothing, iiOpcode = [135], iiRequiredMod = Nothing, iiRequiredReg = Nothing, iiRequiredRM = Nothing}
nop
```

The unusual part is that despite these instructions having different operand sizes (16-bit in the former and 32-bit in the latter), `flexdis86` claims that both instructions have 32-bit (`DWord`) operands. The former mistakenly claims that the 16-bit operands are 32-bit because of `xchg`'s janky treatment in `data/optable.xml`:

https://github.com/GaloisInc/flexdis86/blob/7109bdc9990a3e756eb7fb07419737d15ad41da0/data/optable.xml#L8389-L8393

This has both an operand-size override (`oso`, or `0x66`) as well as an opcode beginning with `0x66`. But the [ISA manual entry for `xchg`](https://www.felixcloutier.com/x86/xchg) shows the opcode being just `0x90`. I suspect that the reason things are this way is to make the opcode table work with the current disassembler machinery, which encodes all prefix combinations and therefore needs the `0x66` bit to disambiguate `xchg` from `nop`, whose opcode is also `0x90`. (See also c20797eaa2b0dc904324262aa5d8ec1bf21125c7, the commit which gave `xchg` this special treatment in the opcode table.)

Besides the bug above, one other strange consequence of this is that `flexdis86` thinks that `66 66 90` is a different instruction that `66 90`:

```
$ cabal run exe:DumpInstr -- 66 90
II {iiLockPrefix = NoLockPrefix, iiAddrSize = Size64, iiOp = "xchg", iiArgs = [(DWordReg eax,OpType (Opcode_reg 0) VSize),(DWordReg eax,OpType (Reg_fixed 0) VSize)], iiPrefixes = Prefixes {_prLockPrefix = NoLockPrefix, _prSP = SegmentPrefix {unwrapSegmentPrefix = 0}, _prREX = 0b00000000, _prVEX = Nothing, _prASO = False, _prOSO = False}, iiRequiredPrefix = Nothing, iiOpcode = [102,144], iiRequiredMod = Nothing, iiRequiredReg = Nothing, iiRequiredRM = Nothing}
nop

$ cabal run exe:DumpInstr -- 66 66 90
II {iiLockPrefix = NoLockPrefix, iiAddrSize = Size64, iiOp = "xchg", iiArgs = [(WordReg ax,OpType (Opcode_reg 0) VSize),(WordReg ax,OpType (Reg_fixed 0) VSize)], iiPrefixes = Prefixes {_prLockPrefix = NoLockPrefix, _prSP = SegmentPrefix {unwrapSegmentPrefix = 0}, _prREX = 0b00000000, _prVEX = Nothing, _prASO = False, _prOSO = True}, iiRequiredPrefix = Nothing, iiOpcode = [102,144], iiRequiredMod = Nothing, iiRequiredReg = Nothing, iiRequiredRM = Nothing}
xchg   ax,ax
```

Moreover, this reveals that there is a special case in the pretty-printer for `xchg ax,ax` that causes it to be printed as `nop`, but there is _not_ a corresponding special case for `xchg eax,eax`. This seems inconsistent.

### PR DESCRIPTION
**Pull request by RyanGlScott**
Previously, the opcode lookup table would encode every possible permutation of allowable prefixes for each instruction as a separate path. This is expensive in both space and time, as observed in #40. The new approach taken in this patch, as described in `Note [x86_64 disassembly]` in `Flexdis86.Disassembler`, is to only encode the VEX prefixe and opcode bytes in the lookup table, leaving out all other forms of prefix bytes entirely. Instead, disassembly will start by eagerly parsing as many prefix bytes as possible, proceeding to parse opcode bytes after the first non-prefix byte is encountered. After identifying the possible instructions from the opcode, we will then narrow down exactly which instruction it is by validating them against the set of parsed prefixes.

As noted in `Note [x86_64 disassembly]`, we had to add some special cases for `nop`-like instructions—namely, `endbr32`, `endbr64`, `pause`, and `xchg`—to avoid some prefix byte–related ambiguity. The new handling for `xchg` is more accurate than it was before, so this patch fixes #42 as a side effect. This patch also addresses part (1) of #40 in that it should reduce the amount of memory that the lookup table uses, although there is potentially more work to be done (see part (2) of #40).

### CODE REVIEW

File: src/Flexdis86/Disassembler.hs [hunk 1]
```diff
@@ -600,17 +742,285 @@ getRMTable modRM mtbl =
...
+                         -> m InstructionInstance
+    disassembleWithModRM prefixBytes defs = do
+      tbl <- checkRequiredReg defs
+      case tbl of
+        RegTable v -> do
+          modRM <- readModRM
+          let mtbl = v V.! fromIntegral (modRM_reg modRM)
+          parseReadTable prefixBytes modRM (getReadTable modRM (getRMTable modRM mtbl))
+        RegUnchecked mtbl -> do
+          modRM <- readModRM
```
Comment (travitch) [inline src/Flexdis86/Disassembler.hs:1014]: Would lifting these `readModRM` calls out of the case be a bit clearer?

File: src/Flexdis86/Disassembler.hs [hunk 2]
```diff
@@ -600,17 +742,285 @@ getRMTable modRM mtbl =
...
+      -- Although we use Data.List.lookup below, this is acceptable since the
+      -- lists being looked up are very small in length.
+      if |  b `elem` simplePrefixBytes
+         -> loopPrefixBytes (prefixBytes Seq.|> b)
+
+         |  b `elem` segPrefixBytes
+         -> loopPrefixBytes (prefixBytes Seq.|> b)
+
+         |  b `elem` rexPrefixBytes
+         -> loopPrefixBytes (prefixBytes Seq.|> b)
```
[the remark this answers is not in the document]
Comment (RyanGlScott) [reply src/Flexdis86/Disassembler.hs]: Alright, you've convinced me. I've switched from lists to `HashSet`s. There were also some other uses of `Data.List.lookup` on lists of pairs that I've converted to HashMap`s.

File: src/Flexdis86/Disassembler.hs (hunk 1 above)
Comment (RyanGlScott) [reply src/Flexdis86/Disassembler.hs]: Sure, why not. Done.

File: src/Flexdis86/Disassembler.hs [hunk 3]
```diff
@@ -600,17 +742,306 @@ getRMTable modRM mtbl =
...
+    loopPrefixBytes prefixBytes = do
+      b <- readByte
+      if |  HS.member b simplePrefixBytes
+         -> loopPrefixBytes (prefixBytes Seq.|> b)
+
+         |  HS.member b segPrefixBytes
+         -> loopPrefixBytes (prefixBytes Seq.|> b)
+
+         |  HS.member b rexPrefixBytes
+         -> loopPrefixBytes (prefixBytes Seq.|> b)
```
[the remark this answers is not in the document]
Comment (RyanGlScott) [reply src/Flexdis86/Disassembler.hs]: I've consolidated all of `simplePrefixBytes`, `segPrefixBytes`, and `rexPrefixBytes` into `nonVexPrefixBytes`.

File: src/Flexdis86/Disassembler.hs [hunk 4]
```diff
@@ -54,6 +63,180 @@ import           Flexdis86.Register
...
+  | Prefix bytes (excluding VEX) | VEX prefix bytes | Opcode bytes | Operand bytes |
+  ----------------------------------------------------------------------------------
+
+One downside to this special case is that we must increase the size of the
+lookup table slightly to accommodate VEX prefix bytes. Without VEX prefix
+bytes, the size of the table would be about 1,400 nodes, but with VEX prefix
+bytes, the size is about 13,000 nodes. That's still several orders of magnitude
+smaller than 5.8 million nodes, however, so we consider this an acceptable
+price to pay.
+
```
[the remark this answers is not in the document]
Comment (RyanGlScott) [reply src/Flexdis86/Disassembler.hs:206]: I've pushed a change which rearranges the prose in this part of the Note a little bit. Can you check to see if this way of phrasing it is more to your liking?

**Review by travitch — APPROVED** [review]

**Comment by RyanGlScott** [comment]
Here is a heap profiling report for the same SAW proof as in https://github.com/GaloisInc/flexdis86/issues/40#issue-1241995531, but with the patch in #43 applied:

[flexdis86-patch.pdf](https://github.com/GaloisInc/flexdis86/files/9030901/flexdis86-patch.pdf)

`mkX64Disassembler` doesn't even show up in this report's top offenders, which is exciting. I think this is as good of a sign as any that this patch does in fact reduce the memory usage in practice.

**Pull request merged**